tensorflow/compiler/xla/service/gpu/gemm_thunk.cc

```diff
@@ -294,81 +411,157 @@ Status RunGemm(const GpuGemmConfig &gemm_config,
+  return std::make_tuple(lhs_matrix, rhs_matrix, out_matrix);
+}
 
+Status RunGemm(const GpuGemmConfig &gemm_config,
+               se::DeviceMemoryBase lhs_buffer, se::DeviceMemoryBase rhs_buffer,
+               se::DeviceMemoryBase output_buffer, se::Stream *stream,
+               bool implements_whole_instruction,
+               absl::optional<int64_t> profile_index,
+               BlasScratchAllocator *scratch_allocator,
+               se::blas::IBlasLtMatmulAlgorithm *const algorithm_being_profiled,
+               se::blas::ProfileResult *profile_result,
+               absl::optional<se::blas::AlgorithmType> algorithm) {
+  VLOG(2) << "Executing a GemmThunk";
+  se::blas::MatrixDescriptor lhs_matrix, rhs_matrix, output_matrix;
+  std::tie(lhs_matrix, rhs_matrix, output_matrix) = PopulateInputOutputMatrices(
+      gemm_config, lhs_buffer, rhs_buffer, output_buffer);
+  const Shape &output_shape = gemm_config.output_shape;
+  int64_t batch_size = gemm_config.backend_config.batch_size();
+  const GemmBackendConfig &backend_config = gemm_config.backend_config;
+  // The BlasLtMatmul routines (unlike BlasGemm, BlasGemmBatched etc.) take
+  // alpha and beta with the same type as the matrices.
   complex128 alpha = {backend_config.alpha_real(), backend_config.alpha_imag()};
   double beta = backend_config.beta();
 
-  switch (output_shape.element_type()) {
-    case S32: {
-      if (!best_algorithm) {
-        return InternalError("Only extended GEMM is supported for int32_t");
+  // The BlasLtMatmul routines are only supported from CUDA 11.0 onward.
+#if defined(GOOGLE_CUDA) && CUDA_VERSION >= 11000
+  absl::flat_hash_set<PrimitiveType> enabled_types = {F16, F32, F64, C64, C128};
+  if (gemm_config.use_cublaslt &&
+      enabled_types.contains(output_shape.element_type())) {
+    bool has_activation_relu =
+        static_cast<se::dnn::ActivationMode>(
+            backend_config.activation_mode()) == se::dnn::ActivationMode::kRelu;
+    se::blas::IBlasLtMatmulAlgorithm *best_algorithm = nullptr;
+    if (algorithm_being_profiled) {
+      best_algorithm = algorithm_being_profiled;
+    }
+    complex128 beta_cmplx = {beta, 0};
+    switch (output_shape.element_type()) {
+      case F16:
+        CHECK_EQ(alpha.imag(), 0);
+        return DoGemmLt<Eigen::half>(
+            batch_size, lhs_matrix, rhs_matrix, output_matrix, stream,
+            static_cast<Eigen::half>(alpha.real()),
+            static_cast<Eigen::half>(beta), scratch_allocator, best_algorithm,
+            /*output_profile_result=*/profile_result);
+      case F32:
+        CHECK_EQ(alpha.imag(), 0);
+        return DoGemmLt<float>(
+            batch_size, lhs_matrix, rhs_matrix, output_matrix, stream,
+            static_cast<float>(alpha.real()), static_cast<float>(beta),
+            scratch_allocator, best_algorithm,
+            /*output_profile_result=*/profile_result);
+      case F64:
```

Comment:
What can we try to do to avoid duplicating the enum->template switch? It occurs in this PR a few times, seems the two in this file can be unified quite easily?

Reply:
++ cf. the comment above.